### ISSUE TITLE
Reject or clamp dates past the year 9999

### ISSUE THREAD

**Issue by codl** [mastodon/mastodon#17673]
### Pitch

Dates returned by the Mastodon API should not go past year 9999.

### Motivation

This is an interoperability issue, 5-digit years have a tendency to break clients (https://github.com/tuskyapp/Tusky/issues/2272, https://github.com/nolanlawson/pinafore/issues/2113).

The format that Mastodon returns for dates with 5-digit years, `YYYYY-MM-DDThh:mm:ss.sssZ` does not conform to ISO 8601, which requires a leading `+` and requires all parties (read: Mastodon and clients) to agree ahead of time on a set number of extra digits.

<details><summary>Snippets from ISO 8601:2004. Warning: really dry standards documentation</summary>

> ## 3.5 Expansion
> 
> By mutual agreement of the partners in information interchange, it is permitted to expand the component
> identifying the calendar year, which is otherwise limited to four digits. This enables reference to dates and
> times in calendar years outside the range supported by complete representations, i.e. before the start of the
> year [0000] or after the end of the year [9999].
> 
> ## 3.6 Leading zeros
> If a time element in a defined representation has a defined length, then leading zeros shall be used as
> required.
> 
> ## 3.7 Mutual agreement
> Some of the representations identified in this International Standard are only allowed by mutual agreement of
> the partners in information interchange. Such agreement should ensure that fields in which the representation
> may occur are not allowed to hold other representations that cannot be unambiguously distinguished from the
> agreed representation.

> #### 4.1.2.4 Expanded representations
> 
> If, by agreement, expanded representations are used, the formats shall be as specified below. The
> interchange parties shall agree the additional number of digits in the time element year. In the examples below
> it has been agreed to expand the time element year with two digits.
> 
>   a) A specific day
>       Basic format: ±YYYYYMMDD Example: +0019850412
>       Extended format: ±YYYYY-MM-DD Example: +001985-04-12
</details>

<details><summary>Snippet from ISO 8601 Wikipedia page. Less dry but non-normative</summary>

> To represent years before 0000 or after 9999, the standard also permits the expansion of the year representation but only by prior agreement between the sender and the receiver. An expanded year representation \[±YYYYY] must have an agreed-upon number of extra year digits beyond the four-digit minimum, and it must be prefixed with a + or − sign

<https://en.wikipedia.org/wiki/ISO_8601#Years>

</details>

Regardless, i've found that support for this expanded representation is spotty among standard libraries or even specialized date-parsing libraries. Some accept only `+YYYYY`, some only accept `YYYYY` only without a leading `+`, some simply do not accept years past 9999 no matter how they are formatted. Hence I believe it would be best to avoid the issue altogether by ensuring that the API does not return dates past year 9999, whether it be by rejecting them at federation time, or by clamping them to our decamillenium at view time, or something else.

While this would set up a Y10K crash scenario, we've got a few more millenia to figure that out. I'm sure by that time ISO 8601 will have been updated and everyone will agree on how to represent five-digit years.

Did you know a period of 10,000 years is called a decamillenium? I didn't until writing this issue.

**Comment by vmstan** [comment]
@ClearlyClaire is this impacted by #27655 at all?

**Comment by ClearlyClaire** [comment]
Yes! #27655 doesn't clamp those values but rejects them, replacing them with current time. I believe this addresses the issue though.
